### PR TITLE
Fix various virtualscroller autoscrolling bugs

### PR DESCRIPTION
**Pull request by adamconroy**
### Intent

Fixes https://github.com/rstudio/rstudio/issues/8544
Fixes https://github.com/rstudio/rstudio/issues/8504
Fixes https://github.com/rstudio/rstudio/issues/8529
Fixes https://github.com/rstudio/rstudio/issues/8552

Wanted to get #8568 also in here but there's some weirdness related to GWT going on 😢 

### Approach

The weird Tibble related bugs were due to how I was counting the number of items in a bucket. A super wide Tibble is a ton of unique `span` elements that can all be on a single line. This caused the calculation of how large a bucket is to be wildly inaccurate so it didn't know if it was scrolled all the way up or down. The `isBucketFull` function in `virtualscroller.js` solves this by just counting newlines in the `innerText` which seems to work pretty well for now, though I did notice a performance hit which might need some attention.

The console always scrolling to the bottom is a...more complex issue. It happens in a lot of places based on a whole bunch of different events and there are a few different `scrollToBottom` insertion points that are hard to grab in a catch all. The `BottomScrollPane` makes a lot of assumptions based on events that happen entirely within GWT and the state doesn't seem to update cleanly between pure JS and the GWT event loop. I refactored out some of the built in autoscrolling and conditionally moved that to virtualscroller.js but there is still more work to be done. This [comment](https://github.com/rstudio/rstudio/issues/8568#issuecomment-745748922) in #8568 shows some of the complexity inherent in having the content and scrolling managed in multiple places.

### QA Notes

The most important thing in testing this will be to ensure that nothing with the Limit Console Output feature turned off is broken. Always good to be a little extra cautious when changing rerouting old code.

### CODE REVIEW

**Comment by kevinushey** [comment]
LGTM!

**Comment by R-Hannibal** [comment]
@adamconroy - Could you provide any more specificity to the QA notes on this PR?

`...ensure that nothing with the Limit Console Output feature turned off is broken...`

I absolutely agree with caution, but if the only way to truly verify this fix is a general regression of IDE functionality, we may want to check the individual tickets and then allow a combination of early adopter usage and certification checks to provide that more general coverage.

Note: Apologies if you meant specific testing around the console with the Limit Output option turned on. I'm reading that as a general regression of the IDE and it's scaring me 😆.

**Comment by kevinushey** [comment]
Testing this is hard because it's hard to quantify all of the ways that users might interact with the Console pane, and the ingrained expectations they have on the way it works. It's only after one of those expectations is violated that the issue starts to stand out.

I would test things like:

- Clicking on the console pane sends focus to console input;
- The console stays scrolled to the bottom when lots of streaming output is emitted;
- The other various glitches reported in the past are ironed out.

Basically, you need to think and behave like a regular R user and see if anything has changed.
